src/kcas.ml

```diff
@@ -10,69 +10,102 @@ module Id = struct
   | CASN (_, state, lt, gt) ->
-      release_after lt |> ignore;
-      state.before <- state.after;
-      state.casn <- casn_after;
-      release_after gt
+      release_after casn lt |> ignore;
+      if not (is_cmp casn state) then (
+        state.before <- state.after;
+        state.casn <- casn_after);
+      release_after casn gt
 
-let rec release_before = function
+let rec release_before casn = function
   | NIL -> false
   | CASN (_, state, lt, gt) ->
-      release_before lt |> ignore;
-      state.after <- state.before;
-      state.casn <- casn_before;
-      release_before gt
-
-(* Note: The writes to `state.casn <- ...` above could be removed to reduce time
-   at the cost of increasing space usage (by a constant factor). *)
-
-let release cass = function
-  | `After -> release_after cass
-  | `Before -> release_before cass
-
-let finish casn (`Undetermined cass as undetermined)
-    (status : [ `Before | `After ]) =
+      release_before casn lt |> ignore;
+      if not (is_cmp casn state) then (
+        state.after <- state.before;
+        state.casn <- casn_before);
+      release_before casn gt
+
+let release casn cass = function
+  | `After -> release_after casn cass
+  | `Before -> release_before casn cass
+
+let rec verify casn = function
+  | NIL -> `After
+  | CASN (atom, desired, lt, gt) -> (
+      match verify casn lt with
+      | `After ->
+          if is_cmp casn desired && Atomic.get atom.state != desired then
+            `Before
+          else verify casn gt
+      | status -> status)
+
+let finish casn (`Undetermined cass as undetermined) (status : determined) =
   if Atomic.compare_and_set casn (undetermined :> status) (status :> status)
-  then release cass status
+  then release casn cass status
   else Atomic.get casn == `After
 
-let rec determine casn undetermined skip = function
-  | NIL -> skip || finish casn undetermined `After
-  | CASN (loc, desired, lt, gt) as eq ->
-      determine casn undetermined true lt
-      &&
-      let current = Atomic.get loc.state in
-      if desired == current then determine casn undetermined skip gt
-      else
-        let expected = desired.before in
-        if
-          expected == current.after
-          && (current.casn == casn_after || is_after current.casn)
-          || expected == current.before
-             && (current.casn == casn_before || not (is_after current.casn))
-        then
-          let status = Atomic.get casn in
-          if status != (undetermined :> status) then status == `After
-          else if Atomic.compare_and_set loc.state current desired then
-            determine casn undetermined skip gt
-          else determine casn undetermined skip eq
-        else finish casn undetermined `Before
+let exit _ = raise Exit [@@inline never]
+
+let rec determine casn action = function
+  | NIL -> action
+  | CASN (loc, desired, lt, gt) as eq -> (
+      match determine casn action lt with
+      | `Before -> `Before
+      | action ->
```

Comment:
I'd vote against using wildcards with polymorphic variants. They can be broken with a typo, and the code will still compile. Personally, I try to avoid them in general, just because it's easier to evolve such a code base. 

```suggestion
      | (`After | `Verify) as action ->
```

Reply:
Note that the `action` value is the return value of the `determine` function and that function is only called from `is_after` and `determine_for_owner`, which both explicitly handle all the cases.  Also, the polymorphic variants ultimately flow into the monotyped `casn` atomic.  So, basically, if you make typo it will be caught.

But it shouldn't hurt to be more explicit in the couple of places you mentioned, so I'll change those spots to be more explicit.